DependencyInjection/Compiler/SwiftMailerPass.php

```diff
@@ -36,7 +36,7 @@ public function process(ContainerBuilder $container)
                 // default class, lets check for Swift Mailer version and set correct class
                 $class = new \ReflectionClass('\\Swift_Spool');
                 $parameters = $class->getMethod('queueMessage')->getParameters();
-                switch ($parameters[0]->getClass()->getName()) {
+                switch ($parameters[0]->getType()) {
```

Comment:
I'm not sure why this change was done, because it seems like it uses __toString of ReflectionType which is deprecated https://www.php.net/manual/en/reflectiontype.tostring.php

Reply:
Interesting... I actually changed it because PHPUnit on PHP 8 reported `ReflectionParameter::getClass()` as deprecated. Unfortunately the failing build is no longer visible after I rebased my commits.
[The stubs in PhpStorm say the same](https://github.com/JetBrains/phpstorm-stubs/blob/ed3c896f6d7fef78dfd42366c8c3fd596676d75a/Reflection/ReflectionParameter.php#L126), but the PHP documentation [doesn't say anything about it](https://www.php.net/manual/en/reflectionparameter.getclass.php).

The PhpStorm stub points to ReflectionParameter::getType() as a replacement:

>     #[Deprecated(reason: "Use ReflectionParameter::getType() and the ReflectionType APIs should be usedinstead.", since: "8.0")]

However I missed the deprecation warning for the `__toString()` method. I'll have to look into this.

Reply:
So it looks like one is deprecated in 8, and one in 7.4, but the one deprecated in 7.4 isn't in 8 anymore?! This is kinda bonkers.

https://3v4l.org/23vV9

Reply:
In any case, it seems to me like this should maybe not belong in this PR to avoid blocking the php 8 support which it would appear works regardless of this change.

Reply:
It seems to be caused by (preparation for) the introduction of union types in PHP 8. `ReflectionType` has become abstract with two classes extending it, [ReflectionNamedType](https://www.php.net/manual/en/class.reflectionnamedtype.php) (since 7.1) and [ReflectionUnionType](https://www.php.net/manual/en/class.reflectionuniontype.php) (since 8.0). Calling `ReflectionParameter::getType()` will return `ReflectionNamedType` if the type hint contains one type, `ReflectionUnionType` if it is a union type, or `null` if there is no type hint. `ReflectionUnionType` has a `getTypes()` method which returns a list of `ReflectionNamedType` instances. `ReflectionNamedType` has a `getName()` method which we are supposed to call instead of relying on `__toString()`.

Unfortunately the documentation on this appears to be incomplete - I will see if I am able to contribute a pull request to php/doc.

As we check the parameter type hint of a known class here (`Swift_Spool`) and expect it to use either `Swift_Mime_Message` or `Swift_Mime_SimpleMessage` as a type hint, I believe it's safe to expect an instance of `ReflectionNamedType`, so I could call `->getName()` on it. That way we don't have to rely on `__toString()`. However, a static analysis tool such as PHPStan would probably fall over this because strictly speaking, `getType()` is not guaranteed to return a `ReflectionNamedType` instance.

If we really want to be on the safe side I could first check if `getType()` returns an instance of `ReflectionNamedType` but as stated above, that should not occur under normal circumstances.